src/tools/tidy/src/bins.rs

```diff
@@ -32,9 +32,12 @@ pub fn check(path: &Path, output: &Path, bad: &mut bool) {
     // readily create a file there to test.
     //
     // See #36706 and #74753 for context.
-    let mut temp_path = path.join("tidy-test-file");
+    //
+    // We also append the thread ID to avoid threads trampling on each others files.
+    let file_name = format!("t{}.tidy-test-file", std::thread::current().id().as_u64());
```

Comment:
This implies we're visiting the same path more than once, which seems surprising to me - why is this necessary?

Reply:
Maybe this isn't actually needed. I did this change before the one in `filter_dirs`. I'll retest with only the `filter_dirs` change.

Reply:
> thread '<unnamed>' panicked at 'Deleted temp file: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/tools/tidy/src/bins.rs:43:46

Looks like adding the thread ID was necessary after all.

Reply:
I'm not satisfied with just an error message - something is wrong with at least my mental model of the code, and that leads me to believe this may just be hiding some other bug. I would like to see an explanation for why the error is happening.

Reply:
One possibility is that the first create fails and then the 2nd attempt uses the output dir which is shared between the passes.

https://github.com/rust-lang/rust/blob/3c10a880eca60379343a6c6d19dd5bda38ead55d/src/tools/tidy/src/bins.rs#L36-L38

Reply:
Can we not parallelize this pass perhaps to avoid this? I'm not convinced by your explanation I guess and it feels subtle; I'd rather have slow correct code than wrong parallel code :)

Reply:
We could, but I don't quite understand the concern. The tests are largely non-mutating. This is the only mutating part and the fix I did would be *necessary* for correctness anyway due to to the possibility mentioned above.

So what do you fear might be going wrong beyond this?

Reply:
The modified comment in the code doesn't suggest a correctness fix to me, so maybe I overlooked that - if there's a correctness fix here, I'm not yet seeing how we've avoided the problem by way of thread ID... does this problem occur before your PR? If so, then I don't see how adding thread ID helps, as previously it'd always be zero, right?

Reply:
What I was trying to say is that it is a *necessary* part of making these checks concurrent because the code path creating the files may create them in either the source or the output folder and the output folder is shared. So if two checks run concurrently then they might create the file twice and try to delete it twice which will fail if it's not a unique file name.

Your concern seems to be that it's not *sufficient*, but I don't see anything else that might trigger this.

Reply:
FWIW (I guess we're moving it to a separate PR) I think I'd suggest we move this out of the parallel code or cache the results; they shouldn't change and doing it just once seems simpler than trying to make sure we don't clobber each other. It also helps us avoid accidentally getting different results in different threads, which could be quite confusing.